onnxruntime/core/providers/cuda/math/softmax_blockwise_impl.cuh

```diff
@@ -0,0 +1,335 @@
+    threadVal = r(threadVal, data[offset]);
+
+  return threadVal;
+}
+
+/**
+ * This will apply the Epilogue with vectorized reads & writes when input & output have the same shift
+ */
+template <int ILP, typename scalar_t, typename accum_t, typename outscalar_t, template<typename, typename, typename> class Epilogue>
+__device__ __forceinline__ void
+WriteFpropResultsVectorized(
+             int size,
+             const int shift,
+             scalar_t *input,
+             outscalar_t *output,
+             Epilogue<scalar_t, accum_t, outscalar_t> epilogue) {
+  using LoadT = aligned_vector<scalar_t, ILP>;
+  using StoreT = aligned_vector<outscalar_t, ILP>;
+
+  int offset = threadIdx.x;
+
+  // if unaligned, do one value / thread and move on, guaranteeing aligned reads/writes later
+  if (shift > 0) {
+    input -= shift;
+    output -= shift;
+    size += shift;
+
+    if (threadIdx.x >= shift) {
+      output[offset] = epilogue(input[offset]);
+    }
+    size -= blockDim.x;
+    input += blockDim.x;
+    output += blockDim.x;
+  }
+
+  const int last = size % (ILP * blockDim.x);
+
+  scalar_t in_v[ILP];
+  LoadT* in_value = reinterpret_cast<LoadT*>(&in_v);
+
+  outscalar_t out_v[ILP];
+  StoreT* out_value = reinterpret_cast<StoreT*>(&out_v);
+
+  for (; offset * ILP < (size - last); offset += blockDim.x) {
+    *in_value = reinterpret_cast<LoadT*>(input)[offset];
+
+    #pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      out_v[j] = epilogue(in_v[j]);
+    }
+
+    reinterpret_cast<StoreT*>(output)[offset] = *out_value;
+  }
+
+  offset = size - last + threadIdx.x;
+  // handle the tail
+  for (; offset < size; offset += blockDim.x) {
+    output[offset] = epilogue(input[offset]);
+  }
+}
+
+
+/**
+ * This will apply the Epilogue with non-vectrorized reads & writes for the general case
+ */
+template <int ILP, typename scalar_t, typename accum_t, typename outscalar_t, template<typename, typename, typename> class Epilogue>
+__device__ __forceinline__ void
+WriteFpropResults(
+             int classes,
+             scalar_t *input,
+             outscalar_t *output,
+             Epilogue<scalar_t, accum_t, outscalar_t> epilogue) {
+  int offset = threadIdx.x;
+
+  int last = classes % (ILP * blockDim.x);
+
+  // Main bulk of loop with ILP
+  for (; offset < classes - last; offset += blockDim.x * ILP) {
+    scalar_t tmp[ILP];
+
+    #pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      tmp[j] = input[offset + j * blockDim.x];
+    }
+    #pragma unroll
+    for (int j = 0; j < ILP; ++j) {
+      output[offset + j * blockDim.x] = epilogue(tmp[j]);
+    }
+  }
+
+  // Remainder - no ILP
+  for (; offset < classes; offset += blockDim.x) {
+    output[offset] = epilogue(input[offset]);
+  }
+}
+
+template <int ILP, typename scalar_t, typename accscalar_t, typename outscalar_t,
+  template <typename, typename, typename> class Epilogue>
+__global__ void
+softmax_block_forward(outscalar_t *output, scalar_t *input, int classes)
```

Comment:
I think we also need one for backward(). 

currently, we only use dispatch_softmax_backward() for D <=1024.

Reply:
You can do it as a separate PR.

Reply:
Yeah I agree.